File: arr/pingfangpaixu.cpp
#include <iostream>
#include <vector>
#include <algorithm>  // sort 排序需要这个头文件
using namespace std;

class Solution {
public:
    vector<int> sortedSquares(vector<int>& A) {
        // 第一步：每个数平方
        for (int i = 0; i < A.size(); i++) {
            A[i] *= A[i];
        }
        // 第二步：排序
        sort(A.begin(), A.end());
        return A;
    }
};

class Solution2 {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        vector<int> result(nums.size(),0);
        int k = result.size()-1;
        int i = 0;
        int j = nums.size()-1;
        while(i<j)
        {
            if(nums[i]*nums[i]<nums[j]*nums[j])
            {
                result[k]=nums[j]*nums[j];
                k--;
                j--;
            }
            else if(nums[i]*nums[i]>=nums[j]*nums[j])
            {
                result[k]=nums[i]*nums[i];
                k--;
                i++;
            }
        }
        result[k] = nums[i]*nums[i];
        return result;
    }
};

// 打印数组
void printVector(vector<int>& v) {
    for (int i = 0; i < v.size(); i++) {
        cout << v[i] << " ";
    }
    cout << endl;
}

int main() {
    Solution sol;

    // 测试用例：[-4,-1,0,3,10]
    vector<int> nums = {-4, -1, 0, 3, 10};
    
    cout << "原数组：";
    printVector(nums);

    vector<int> res = sol.sortedSquares(nums);

    cout << "平方+排序后：";
    printVector(res);

    return 0;
}

File: arr/pingfangpaixu.java
import java.util.Arrays;  


public class pingfangpaixu {

    // 主函数：程序入口
    public static void main(String[] args) {
        Solutionpfpx solution = new Solutionpfpx();

        // 测试用例
        int[] nums = {-4, -1, 0, 3, 10};

        // 方法1：暴力平方 + 排序
        int[] res1 = solution.sortedSquares(nums.clone()); // 用clone避免原数组被修改
        System.out.println("方法1（平方+排序）：" + Arrays.toString(res1));

        // 方法2：双指针最优解法
        int[] res2 = solution.sortedSquares2(nums);
        System.out.println("方法2（双指针）：" + Arrays.toString(res2));
    }
}


class Solutionpfpx {
    // 方法1：暴力法 平方 + 排序
    public int[] sortedSquares(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            nums[i] = nums[i] * nums[i];
        }
        Arrays.sort(nums);
        return nums;
    }

    // 方法2：双指针法（最优 O(n)）
    public int[] sortedSquares2(int[] nums) {
        int right = nums.length - 1;
        int left = 0;
        int[] result = new int[nums.length];
        int index = result.length - 1;
        
        while (left <= right) {
            if (nums[left] * nums[left] > nums[right] * nums[right]) {
                result[index--] = nums[left] * nums[left];
                left++;
            } else {
                result[index--] = nums[right] * nums[right];
                right--;
            }
        }
        return result;
    }
}